src/jit/MemoryInl.h

```diff
@@ -1011,3 +1011,326 @@ static void emitStore(sljit_compiler* compiler, Instruction* instr)
+    *result = shared->fetch_or(*value);
+}
+
+static void atomicRmwXor64(uint64_t* shared_p, int64_t* value, int64_t* result)
+{
+    std::atomic<uint64_t>* shared = reinterpret_cast<std::atomic<uint64_t>*>(shared_p);
+    *result = shared->fetch_xor(*value);
+}
+
+static void atomicRmwXchg64(uint64_t* shared_p, int64_t* value, int64_t* result)
+{
+    std::atomic<uint64_t>* shared = reinterpret_cast<std::atomic<uint64_t>*>(shared_p);
+    *result = shared->exchange(*value);
+}
+
+static void emitAtomicRmw64(sljit_compiler* compiler, Instruction* instr)
+{
+    uint32_t options = MemAddress::CheckNaturalAlignment | MemAddress::AbsoluteAddress;
+    uint32_t size = 8;
+    sljit_u32 offset;
+
+    Operand* operands = instr->operands();
+    MemAddress addr(options, instr->requiredReg(0), instr->requiredReg(1), 0);
+    AtomicRmw* rmwOperation = reinterpret_cast<AtomicRmw*>(instr->byteCode());
+    offset = rmwOperation->offset();
+
+    addr.check(compiler, operands, offset, size);
+
+    if (addr.memArg.arg == 0) {
+        return;
+    }
+
+    JITArgPair srcArgPair(operands + 1);
+    JITArgPair dstArgPair(operands + 2);
+    sljit_s32 type = SLJIT_ARGS3V(P, P, P);
+    sljit_s32 faddr;
+
+    switch (instr->opcode()) {
+    case ByteCode::I64AtomicRmwAddOpcode: {
+        faddr = GET_FUNC_ADDR(sljit_sw, atomicRmwAdd64);
+        break;
+    }
+    case ByteCode::I64AtomicRmwSubOpcode: {
+        faddr = GET_FUNC_ADDR(sljit_sw, atomicRmwSub64);
+        break;
+    }
+    case ByteCode::I64AtomicRmwAndOpcode: {
+        faddr = GET_FUNC_ADDR(sljit_sw, atomicRmwAnd64);
+        break;
+    }
+    case ByteCode::I64AtomicRmwOrOpcode: {
+        faddr = GET_FUNC_ADDR(sljit_sw, atomicRmwOr64);
+        break;
+    }
+    case ByteCode::I64AtomicRmwXorOpcode: {
+        faddr = GET_FUNC_ADDR(sljit_sw, atomicRmwXor64);
+        break;
+    }
+    case ByteCode::I64AtomicRmwXchgOpcode: {
+        faddr = GET_FUNC_ADDR(sljit_sw, atomicRmwXchg64);
+        break;
+    }
+    default: {
+        RELEASE_ASSERT_NOT_REACHED();
+        break;
+    }
+    }
+
+    if (srcArgPair.arg1 != SLJIT_MEM1(kFrameReg)) {
+        sljit_emit_op1(compiler, SLJIT_MOV, SLJIT_MEM1(kContextReg), OffsetOfContextField(tmp1) + WORD_LOW_OFFSET, srcArgPair.arg1, srcArgPair.arg1w);
+        sljit_emit_op1(compiler, SLJIT_MOV, SLJIT_MEM1(kContextReg), OffsetOfContextField(tmp1) + WORD_HIGH_OFFSET, srcArgPair.arg2, srcArgPair.arg2w);
+    }
+
+    if (srcArgPair.arg1 == SLJIT_MEM1(kFrameReg)) {
+        sljit_emit_op2(compiler, SLJIT_ADD, SLJIT_R1, 0, kFrameReg, 0, SLJIT_IMM, srcArgPair.arg1w - WORD_LOW_OFFSET);
+    } else {
+        ASSERT(SLJIT_IS_REG(srcArgPair.arg1) || SLJIT_IS_IMM(srcArgPair.arg1));
+        sljit_emit_op2(compiler, SLJIT_ADD, SLJIT_R1, 0, kContextReg, 0, SLJIT_IMM, OffsetOfContextField(tmp1));
+    }
+
+    sljit_emit_op1(compiler, SLJIT_MOV, SLJIT_R0, 0, GET_SOURCE_REG(addr.memArg.arg, instr->requiredReg(0)), 0);
```

Comment:
addr.memArg.arg must be a register, the second argument is not needed. Furthermore, if `addr.memArg.arg` is `SLJIT_R1`, the previous code overwrites it.